sql/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/SparkConnectPlanExecution.scala

```diff
@@ -141,19 +143,48 @@ private[execution] class SparkConnectPlanExecution(executeHolder: ExecuteHolder)
 
     var numSent = 0
     def sendBatch(bytes: Array[Byte], count: Long, startOffset: Long): Unit = {
-      val response = proto.ExecutePlanResponse
-        .newBuilder()
-        .setSessionId(sessionId)
-        .setServerSideSessionId(sessionHolder.serverSessionId)
+      if (isResultChunkingEnabled) {
+        // Result chunking is enabled. Split the result batch into multiple chunks if needed.
+        val maxChunkSize = spark.conf.get(CONNECT_SESSION_RESULT_CHUNKING_MAX_CHUNK_SIZE).toInt
+        val numChunks = (bytes.length + maxChunkSize - 1) / maxChunkSize
 
-      val batch = proto.ExecutePlanResponse.ArrowBatch
-        .newBuilder()
-        .setRowCount(count)
-        .setData(ByteString.copyFrom(bytes))
-        .setStartOffset(startOffset)
-        .build()
-      response.setArrowBatch(batch)
-      responseObserver.onNext(response.build())
+        for (i <- 0 until numChunks) {
```

Comment:
Please don't use for comprehensions... Either use an actual loop, or the functional equivalent this translates into.